### PR TITLE
go/sgx/pcs: GetPCKCertificateChain client and minor fixes

### PR DESCRIPTION
**Pull request by ptrus**
As part of work on supporting PPID PCS attestation the `GetPCKCertificateChain` http client was implemented and some minor fixes were done to the existing code.

Although the PPID PCS attestation remains unsupported (due to aesmd api limitations) it makes sense getting this in for potential future implementation.

### CODE REVIEW

File: go/runtime/host/sgx/ecdsa.go [hunk 1]
```diff
@@ -114,6 +114,10 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
 		// We have a PPID, need to retrieve PCK certificate first.
 		// TODO: Fetch PCK certificate based on PPID and include it in the quote, replacing the
 		//       PPID certification data with the PCK certificate chain certification data.
+		//       e.g. sp.pcs.GetPCKCertificateChain(ctx, nil, data.PPID, data.CPUSVN, data.PCESVN, data.PCEID)
+		//
+		//	 Due to aesmd QuoteEx APIs not supporting certification data this currently
+		//       cannot be easily implemented. Instead we rely on a quote provider to be installed.
 		return nil, fmt.Errorf("PPID certification data not yet supported")
```
Comment (kostko) [inline go/runtime/host/sgx/ecdsa.go:121]: Maybe change the error message to say something about the quote provider not being installed?

File: go/common/sgx/pcs/quote.go [hunk 2]
```diff
@@ -456,6 +457,17 @@ func (qs *QuoteSignatureECDSA_P256) Verify(
 		return err
 	}
 
+	// Ensure PCK certification information matches the quote.
```
Comment (kostko) [inline go/common/sgx/pcs/quote.go:460]: Could these changes break consensus? Or would this just fail with a different error below?

Comment (ptrus) [reply go/common/sgx/pcs/quote.go:460]: I think that the `qs.QESignature.Verify` below should fail unless something weird was going on. But we can skip this change for now to be extra sure.

File: go/runtime/host/sgx/ecdsa.go (hunk 1 above)
Comment (ptrus) [reply go/runtime/host/sgx/ecdsa.go:121]: Yeah good idea

File: go/common/sgx/pcs/quote.go [hunk 3]
```diff
@@ -456,6 +457,17 @@ func (qs *QuoteSignatureECDSA_P256) Verify(
...
+	if !bytes.Equal(pckInfo.CPUSVN[:], qs.QEReport.CPUSVN[:]) {
+		return fmt.Errorf("pcs/quote: PCK CPUSVN: '%v', does not match QE report CPUSVN: '%v'",
+			hex.EncodeToString(pckInfo.CPUSVN[:]),
+			hex.EncodeToString(qs.QEReport.CPUSVN[:]),
+		)
+	}
+	if pckInfo.PCESVN != int32(header.PCESVN) {
```
Comment (peternose) [inline go/common/sgx/pcs/quote.go:467]: We should probably change `pcesvn` from `int32` to `int16` everywhere.

File: go/common/sgx/pcs/http.go [hunk 4]
```diff
@@ -133,6 +136,80 @@ func (hc *httpClient) GetTCBBundle(ctx context.Context, fmspc []byte) (*TCBBundl
...
+			PCESVN           string `json:"pcesvn"`
+			PCEID            string `json:"pceid"`
+		}{
+			PlatformManifest: hex.EncodeToString(platformData),
+			CPUSVN:           hex.EncodeToString(cpusvn[:]),
+			PCESVN:           hex.EncodeToString(pcesvnBytes[:]),
+			PCEID:            hex.EncodeToString(pceidBytes[:]),
+		})
+		if merr != nil {
+			return nil, fmt.Errorf("pcs: failed to marshal PCK certificate request payload: %w", err)
```
Comment (peternose) [inline go/common/sgx/pcs/http.go:176]:
```suggestion
			return nil, fmt.Errorf("pcs: failed to marshal PCK certificate request payload: %w", merr)
```

File: go/common/sgx/pcs/quote.go (hunk 3 above)
Comment (ptrus) [reply go/common/sgx/pcs/quote.go:467]: Sure ok. Likely non breaking because a (malovent) non uint16 `pcesvn` would fail somewhere in the validation chain now already.

File: go/common/sgx/pcs/quote.go (hunk 2 above)
Comment (kostko) [reply go/common/sgx/pcs/quote.go:466]: Let's skip these checks for now.

Comment (peternose) [reply go/common/sgx/pcs/quote.go:466]: Or you can only print the error.

Comment (kostko) [reply go/common/sgx/pcs/quote.go:466]: I would avoid pulling in the logger here. Also it seems this is actually making the unit tests fail so I would assume this check is not entirely per spec.

File: go/common/sgx/pcs/http.go [hunk 5]
```diff
@@ -133,6 +136,80 @@ func (hc *httpClient) GetTCBBundle(ctx context.Context, fmspc []byte) (*TCBBundl
...
+		}{
+			PlatformManifest: hex.EncodeToString(platformData),
+			CPUSVN:           hex.EncodeToString(cpusvn[:]),
+			PCESVN:           hex.EncodeToString(pcesvnBytes[:]),
+			PCEID:            hex.EncodeToString(pceidBytes[:]),
+		})
+		if merr != nil {
+			return nil, fmt.Errorf("pcs: failed to marshal PCK certificate request payload: %w", merr)
+		}
+		rsp, err = hc.doPCSRequest(ctx, u, http.MethodPost, "application/json", bytes.NewReader(payload), false) // nolint: bodyclose
```
Comment (peternose) [inline go/common/sgx/pcs/http.go:178]: Not part of this PR, but I think that `doPCSRequest` should call `rsp.Body.Close()` if response status is not OK (same holds for `doIASRequest`).

**Review by kostko — APPROVED** [review]

**Review by peternose — APPROVED** [review]
Code looks good, but I did not test GET and POST requests. Did you?

File: go/common/sgx/pcs/http.go (hunk 5 above)
Comment (ptrus) [reply go/common/sgx/pcs/http.go:179]: Good notice, updated :+1:

**Review by peternose — APPROVED** [review]

**Comment by ptrus** [comment]
> Code looks good, but I did not test GET and POST requests. Did you?

Yes did, as part of trying to get the PPID PCS attestation to work, which in the end didn't work (without installing a quote provider) due to a unrelated issue.
